driver-core/src/main/com/mongodb/client/model/expressions/ExpressionCodecProvider.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import com.mongodb.annotations.Immutable;
+import org.bson.codecs.Codec;
+import org.bson.codecs.configuration.CodecProvider;
+import org.bson.codecs.configuration.CodecRegistry;
+
+@Immutable
+public class ExpressionCodecProvider implements CodecProvider {
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> Codec<T> get(final Class<T> clazz, final CodecRegistry registry) {
+        // TODO: should this be MqlExpression.class?
+        if (Expression.class.isAssignableFrom(clazz)) {
+            return (Codec<T>) new ExpressionCodec(registry);
```

Comment:
I commented on this in the ExpressionCodec.  I think this should be `MqlExpression` if that's what we're able to encode.  No point in saying we can encode `Expression` if it's just going to throw a `ClassCastException` when you try.

Reply:
> I commented on this in the ExpressionCodec. I think this should be MqlExpression if that's what we're able to encode. No point in saying we can encode Expression if it's just going to throw a ClassCastException when you try @jyemin 

I partially implemented this and then balked. We have `class MqlExpression<T extends Expression>`, but the T is just a "fake" <Expression> for most Expressions (it's there for arrays, which are the only container expression type). I get "raw use" for `MqlExpressionCodec implements Codec<MqlExpression>`, but it seems odd that there should be a T in there for arbitrary MqlExpressions.

I'm not very sure, but I think perhaps we really are trying to encode Expressions into BSON. It is an implementation detail that the only class that implements Expression is MqlExpression. (And consequently, there would never be a ClassCastException). If we create other implementations, then it seems we should handle them as part of this codec.

Reply:
> I'm not very sure, but I think perhaps we really are trying to encode Expressions into BSON

Perhaps, but it's not possible since an `Expression` isn't rich enough as a type to be able to build a generic Codec for.  

A CodecProvider should only provide a codec for a type that it is able to encode.  In this case, that type is MqlExpression.  So I think this should just be `MqlExpressionCodecProvider` and it should return instances of `MqlExpressionCodec, which should implement `Codec<MqlExpression>`.

If we end up adding more Expression implementations, we can add a new CodecProvider for those implementations.